### ISSUE TITLE
deprecate PHP3Script and PHPScript in twisted.web.twcgi, replace them with documentation for FilteredScript

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7275]
|<img alt="Jerub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Jerub reported|
|-|-|
|Trac ID|trac#516|
|Type|defect|
|Created|2004-02-20 10:48:30Z|
|Branch|https://github.com/twisted/twisted/tree/phpscript-deprecation-516|

Attachments:

* [patch-twcgi](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b90/b903ea6300ffc957d946618424318ce4afcc0764/1e2821eac97b680ec5a9943f71663aec2d706041) (823 bytes) - added by Jerub on 2004-02-20 10:59:12Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__516 516
type__defect defect
reporter__Jerub Jerub
priority__high high
milestone__ 
branch__branches_phpscript_deprecation_516 branches/phpscript-deprecation-516
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1077274110000000 1077274110000000
changetime__1267288402000000 1267288402000000
version__ 
owner__ 
cc__radix cc__moshez cc__Jerub cc__exarkun cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Jerub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Jerub commented|
|-|-|

```
#!html
<pre>
/usr/bin/php4 in debian unstable is the 'cli' version, a misguided product of a
push to see php in areas such as the command line.

the location of the cgi binary is in /usr/lib/cgi-bin/php4 - and in order for
the CGI version to function correctly, the env variable 'REDIRECT_STATUS' must
be '200'.

This patch probably isn't *optimal*, and will likely break other deployments
where /usr/bin/php4 is the cgi version and the other path doesn't actually
exist, but it demonstrates exactly what is required at a code level for this to
function on debian unstable.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
I suggest this patch get applied to debian packaged versions of Twisted when
they get generated.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

I suppose it would be good to have support for PHP4 CGIs on Debian.  Did anyone contact Debian to discuss this issue?  Also, couldn't we check for the existence of various executables instead of just assuming /usr/bin/php4 exists and is usable?

In any case, this isn't high priority, and the change needs test coverage.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

`PHP3Script` and `PHPScript` are _nothing_ except platform-specific configuration.  If we can't get this right, we shouldn't have it.  We can have documentation which explains how to trivially create one of these which is correct for a particular location of the php cgi interpreter.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @jesstess|
|-|-|

This fruit hangs low.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess commented|
|-|-|

(In [28136]) Branching to 'phpscript-deprecation-516'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess commented|
|-|-|

(In [28146]) Deprecate PHPScript and PHP3Script and add unit tests for the deprecation.

refs [#7275](https://github.com/twisted/twisted/issues/7275)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess removed owner|
|-|-|

I also added rather mundane docstrings for FilteredScript.runProcess and CGIScript.runProcess; more interesting descriptions are welcome.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @jesstess|
|-|-|

Thanks!  The docstring additions look good.  A couple other things:

  * There's one pyflakes warning for the module, an unused local in `CGIScript.render`.  Can you fix that?
  * I think a brief section in the prose documentation for `CGIScript` and `FilteredScript` would be a good addition as well.  I want to suggest adding it to using-twistedweb.xhtml, but it also occurs to me that it could be a section in the new _sixty seconds_ documentation except that that branch hasn't been merged yet.  Blocking this ticket on the _sixty seconds_ ticket sounds lame.  I do want to push us towards writing more prose documentation, though.  Maybe I should do a _sixty seconds_ post on twcgi and we can add it to the _sixty seconds_ branch (or to another ticket after the _sixty seconds_ branch is merged).  (Sorry for the rambling non-committal review point.)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess commented|
|-|-|

(In [28172]) Remove unused local variable and clean up surrounding doc strings.

refs [#7275](https://github.com/twisted/twisted/issues/7275)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess removed owner|
|-|-|

Ticket [#4273](https://github.com/twisted/twisted/issues/4273) was opened to address exarkun's second review comment.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @jesstess|
|-|-|

Oops, I missed some test failures before.  Looks like the new `__init__` methods aren't compatible with the inherited `__init__`:

```
===============================================================================
[ERROR]: twisted.web.test.test_tap.ServiceTests.test_php3Processor

Traceback (most recent call last):
  File "/srv/bb-slave/Run/slave/reactors/Twisted/twisted/web/test/test_tap.py", line 80, in test_php3Processor
    self.assertIsInstance(root.getChild("foo.php3", None), PHP3Script)
  File "/srv/bb-slave/Run/slave/reactors/Twisted/twisted/web/static.py", line 311, in getChild
    return resource.IResource(processor(fpath.path, self.registry))
exceptions.TypeError: __init__() takes exactly 1 argument (3 given)
===============================================================================
[ERROR]: twisted.web.test.test_tap.ServiceTests.test_phpProcessor

Traceback (most recent call last):
  File "/srv/bb-slave/Run/slave/reactors/Twisted/twisted/web/test/test_tap.py", line 90, in test_phpProcessor
    self.assertIsInstance(root.getChild("foo.php", None), PHPScript)
  File "/srv/bb-slave/Run/slave/reactors/Twisted/twisted/web/static.py", line 311, in getChild
    return resource.IResource(processor(fpath.path, self.registry))
exceptions.TypeError: __init__() takes exactly 1 argument (3 given)
-------------------------------------------------------------------------------
```

One other thing.  In the class docstring for `FilteredScript`, I think the sentence about customizing `filter` can be a bit broader - ie, this works for things other than PHP. :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_13"></a>|@jesstess commented|
|-|-|

(In [28177]) Make tap tests that check warnings be aware of the new PHPScript and
PHP3Script warnings.

refs [#7275](https://github.com/twisted/twisted/issues/7275)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess commented|
|-|-|

(In [28179]) FilteredScript is for more than PHP.

refs [#7275](https://github.com/twisted/twisted/issues/7275)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Hmm.  I'm wondering about the warnings that get emitted by `twistd -n web --path ...` now.  It'd be nice to avoid those.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra set owner to @jesstess|
|-|-|

See previous comment.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_18"></a>|@jesstess commented|
|-|-|

(In [28380]) Delay calling PHP3Script and PHPScript in tap.Options.opt_path to
avoid displaying deprecation warnings on something like twistd -n web
--path .

refs [#7275](https://github.com/twisted/twisted/issues/7275)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @jesstess|
|-|-|

  1. test_tap.py has its copyright date bumped, but no other changes.  This should be reverted.
  1. `test_PHP3ScriptIsDeprecated` and `test_PHPScriptIsDeprecated` use `getattr`.  Instead, they can just do `twcgi.PHP3Script` and `twcgi.PHPScript`.
  1. 10.0 has sailed.  Update the deprecation version numbers to 10.1.
  1. As long as you're reformatting the `CGIScript` class docstring, can you switch it away from the _I ..._ style?  Something like _L{CGIScript} is a resource which runs child processes according to the CGI specification._ would be nice.

That's all.  Please merge when the above points are addressed.  Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_21"></a>|@jesstess commented|
|-|-|

(In [28558]) Don't use getattr; we're up to 10.1.

refs [#7275](https://github.com/twisted/twisted/issues/7275)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_22"></a>|@jesstess set status to `closed`|
|-|-|

(In [28559]) Merge phpscript-deprecation-516

Author: jesstess
Reviewer: exarkun
Fixes: [#7275](https://github.com/twisted/twisted/issues/7275)

twisted.web.twcgi.PHP3Script and twisted.web.twcgi.PHPScript are now
deprecated in favor of twisted.web.twcgi.FilterScript.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
